src/Microsoft.ML/Data/TextLoader.cs

```diff
@@ -0,0 +1,152 @@
+            bool supportSparse = true, bool trimWhitespace = false)
+        {
+            var fields = typeof(TInput).GetFields();
+            Arguments.Column = new TextLoaderColumn[fields.Length];
+            for (int index = 0; index < fields.Length; index++)
+            {
+                var field = fields[index];
+                var mappingAttr = field.GetCustomAttribute<ColumnAttribute>();
+                if (mappingAttr == null)
+                    throw Contracts.Except($"{field.Name} is missing ColumnAttribute");
+
+                if (Regex.Match(mappingAttr.Ordinal, @"[^(0-9,\*\-~)]+").Success)
+                    throw Contracts.Except($"{mappingAttr.Ordinal} contains invalid characters. " +
+                        $"Valid characters are 0-9, *, - and ~");
+
+                var name = mappingAttr.Name ?? field.Name;
+                if (name.Any(c => !Char.IsLetterOrDigit(c)))
+                    throw Contracts.Except($"{name} is not alphanumeric.");
+
+                Runtime.Data.TextLoader.Range[] sources;
+                if (!Runtime.Data.TextLoader.Column.TryParseSourceEx(mappingAttr.Ordinal, out sources))
+                    throw Contracts.Except($"{mappingAttr.Ordinal} could not be parsed.");
+
+                Contracts.Assert(sources != null);
+
+                TextLoaderColumn tlc = new TextLoaderColumn();
+                tlc.Name = name;
+                tlc.Source = new TextLoaderRange[sources.Length];
+                DataKind dk;
+                if (!TryGetDataKind(field.FieldType.IsArray ? field.FieldType.GetElementType() : field.FieldType, out dk))
+                    throw Contracts.Except($"{name} is of unsupported type.");
+
+                tlc.Type = dk;
+
+                for (int indexLocal = 0; indexLocal < tlc.Source.Length; indexLocal++)
+                {
+                    tlc.Source[indexLocal] = new TextLoaderRange
+                    {
+                        AllOther = sources[indexLocal].AllOther,
+                        AutoEnd = sources[indexLocal].AutoEnd,
+                        ForceVector = sources[indexLocal].ForceVector,
+                        VariableEnd = sources[indexLocal].VariableEnd,
+                        Max = sources[indexLocal].Max,
+                        Min = sources[indexLocal].Min
+                    };
+                }
+
+                Arguments.Column[index] = tlc;
+            }
+
+            Arguments.HasHeader = useHeader;
+            Arguments.Separator = new[] { separator };
+            Arguments.AllowQuoting = allowQuotedStrings;
+            Arguments.AllowSparse = supportSparse;
+            Arguments.TrimWhitespace = trimWhitespace;
+
+            return this;
+        }
+
+        /// <summary>
+        /// Try to map a System.Type to a corresponding DataKind value.
+        /// </summary>
+        private static bool TryGetDataKind(Type type, out DataKind kind)
+        {
+            Contracts.CheckValueOrNull(type);
```

Comment:
I consider this somewhat confusing, since by declaring this you're saying that the input type *might* be `null`, but as near as I can see from the one usage of this function, it cannot be. So why was it structured this way? Copy-paste, I guess, a deeper reason I don't appreciate?

If there is no other reason maybe make it assert (not check, since it's private) that it can't be null.